lumicks/pylake/kymotracker/kymotrack.py

```diff
@@ -703,25 +713,84 @@ class KymoTrackGroup:
 
     def __init__(self, kymo_tracks):
         self._src = kymo_tracks
-        if self:
-            if len(set(kymo_tracks)) != len(kymo_tracks):
-                raise ValueError(
-                    "Some tracks appear multiple times. The provided tracks must be unique."
-                )
+        self._kymos = self._validate_compatible_sources()
+
+    def _validate_compatible_sources(self, additional_tracks=()):
+        """Check that source kymos for all tracks (including in self) are compatible.
 
-            self._validate_single_source(kymo_tracks)
+        Parameters
+        ----------
+        additional_tracks : KymoTrackGroup, optional
+            Additional tracks to be added to the current instance.
 
-    def _validate_single_source(self, kymo_tracks):
-        kymos = set([track._kymo._id for track in kymo_tracks])
-        channels = set([track._channel for track in kymo_tracks])
+        Returns
+        -------
+        set
+            Set of source kymograph instances
+        """
+        tracks = list(itertools.chain(self, additional_tracks))
+        if not len(tracks):
+            return set()
 
-        if len(kymos) > 1:
-            raise ValueError("All tracks must have the same source kymograph.")
+        if len(set(tracks)) != len(tracks):
+            raise ValueError(
+                "Cannot extend this KymoTrackGroup with a KymoTrack that is already part of the group"
+                if additional_tracks
+                else "Some tracks appear multiple times. The provided tracks must be unique."
+            )
 
+        channels = {track._channel for track in tracks}
         if len(channels) > 1:
             raise ValueError("All tracks must be from the same color channel.")
 
-        return next(iter(kymos)), next(iter(channels))
+        kymos = {track._kymo for track in tracks}
```

Comment:
```suggestion
        kymos = {track._kymo: None for track in tracks}
```

Just a thought, but it might be worthwhile to enforce insertion order to prevent surprises.

We could use a dummy dictionary as a poor man's insertion-ordered set. For most intents and purposes it'll work the same, but we get a guaranteed insertion order, which might prevent surprises. A concrete example, when working with a non-identifiable model, and pulling out the tracks in set order, I noticed that results were different (the order of the residual seems to matter).

Reply:
hmm, seems a bit odd that the order of the residual matters? but we can do this, like you said just to avoid any surprises. I think I came up with a pretty compact yet readable way to basically make a `OrderedTupleFromDictFromSet` 😆 that's surprisingly still pretty readable. Added benefit: we can get rid of all the `next(iter(set))` nonsense

Reply:
Great!